packages/massa-proto-cli/src/as-gen.ts

```diff
@@ -0,0 +1,135 @@
+import { writeFileSync } from 'fs';
+import { spawnSync } from 'child_process';
+import path from 'path';
+import { ProtoFile } from './protobuf';
+
+/**
+ * Converts the given protobuf type to the related assembly script type.
+ *
+ * @param type - the protobuf type
+ *
+ * @returns the assembly script type
+ */
+function convertTypeToAS(type: string): string {
+  switch (type) {
+    case 'bool':
+      return 'bool';
+    case 'int32':
+      return 'i32';
+    case 'int64':
+      return 'i64';
+    case 'uint32':
+      return 'u32';
+    case 'uint64':
+      return 'u64';
+    case 'float':
+      return 'f32';
+    case 'double':
+      return 'f64';
+    case 'string':
+      return 'string';
+    default:
+      throw new Error(`Unsupported type: ${type}`);
+  }
+}
+
+/**
+ * Creates imports for the caller file generated by {@link generateAsCall}
+ *
+ * @param protoData - the proto file data
+ *
+ * @returns the multiline import statements as a string
+ */
+function generateAsImports(protoData: ProtoFile): string {
+  let imports = `import { encode${protoData.funcName}Helper, ${protoData.funcName}Helper } `;
+  imports += `from './${protoData.funcName}Helper';\n`;
+
+  if (protoData.resType !== null) {
+    imports += `import { decode${protoData.funcName}RHelper, ${protoData.funcName}RHelper }`;
+    imports += `from './${protoData.funcName}RHelper';\n`;
+  }
+
+  imports += 'import { call } from "@massalabs/massa-as-sdk";\n\n';
+
+  return imports;
+}
+
+/**
+ * Creates a contract function caller with the given proto file and address.
+ *
+ * @param protoData - the proto file containing the informations to call the contract's function
+ * @param address - the address of the contract containing the function to call
+ * @param outputDirectory - the output directory to create the file for the caller
+ */
+function generateAsCall(
+  protoData: ProtoFile,
+  address: string,
+  outputDirectory: string,
+) {
+  // generating AS arguments
+  let args: string[] = [];
+  protoData.argFields.forEach(({ name, type }) =>
+    args.push(`${name}: ${convertTypeToAS(type)}`),
+  );
+
+  // generating caller
+  let caller = `export function ${protoData.funcName}(${
+    args.length > 0 ? args.join(', ') + ', coins: number' : 'coins: number'
+  }) : ${protoData.resType !== null ? protoData.resType : 'void'} {\n`;
+  caller += `  const result = call("${address}", "${
+    protoData.funcName
+  }", changetype<StaticArray<u8>>(encode${protoData.funcName}Helper(new ${
+    protoData.funcName
+  }Helper(${Array.from(protoData.argFields.keys()).join(', ')}))), coins);\n`;
+
+  if (protoData.resType !== null) {
+    caller += `  const response = decode${protoData.funcName}RHelper(`;
+    caller += `Uint8Array.wrap(changeType<ArrayBuffer>(result)));\n\n`;
+    caller += '  return response.value;\n';
+  }
+  caller += '}';
+
+  writeFileSync(
+    generateAsImports(protoData) + caller,
+    path.join(outputDirectory, protoData.funcName + '.ts'),
+  );
```

Comment:
Clarifying the code by:
- splitting clearly the function code generation in three parts: signature, body and response
- fix bug in writeFileSync arg order
- apply same strategies as defned above
```suggestion
  // Generate function signature
  const functionSignature = `export function ${protoData.funcName}(${
    args.length > 0 ? args.join(', ') + ', ' : ""} coins: number): ${protoData.resType !== null ? protoData.resType : 'void'}`;

  // Generate function body
  const functionBody = `const result = call(
    "${address}",
    "${protoData.funcName}",
    changetype<StaticArray<u8>>(encode${protoData.funcName}Helper(new ${protoData.funcName}Helper(${args.join(', ')}))),
    coins);`;

  let responseDecoding = "";
  if (protoData.resType !== null) {
    responseDecoding = `const response = decode${protoData.funcName}RHelper(Uint8Array.wrap(changeType<ArrayBuffer>(result)));

  return response.value;`;


  // Compose the full function
  const fullFunction = `${functionSignature} {
  ${functionBody}

  ${responseDecoding}
}`;

  // Write to file
  writeFileSync(
    path.join(outputDirectory, `${protoData.funcName}.ts`),
    generateAsImports(protoData) + fullFunction,
  );
}
```